docs/2_2_1_fingerprinting.md

```diff
@@ -65,116 +65,225 @@ It is used to uniquely match a method by its characteristics.
-package com.some.app.ads;
+    invoke-direct {p0}, Lcom/some/app/ads/AdsLoader;->showBannerAds()V
 
-<accessFlags>
+    # Filter 5 target instruction.
+    :cond_16
+    const/16 p1, 0x539
 
-class AdsLoader {
-    public final boolean <methodName>(boolean <parameter>)
+    # Filter 6 target instruction.
+    if-eq p2, p1, :cond_1c
+
+    const/4 p1, 0x1
+
+    goto :goto_1d
 
-    {
-        // ...
+    :cond_1c
+    const/4 p1, 0x0
 
-        var userStatus = "pro";
+    :goto_1d
+    return p1
+.end method
+```
 
-        // ...
+## ⛳️ Example fingerprint
 
-        return <returnValue >;
+```kt
+val hideAdsFingerprint by fingerprint {
+    // Method signature:
+    accessFlags(AccessFlags.PUBLIC, AccessFlags.FINAL)
+    returns("Z")
+    // Last parameter is simply `L` since it's an obfuscated class.
+    parameters("Ljava/lang/String;", "I", "L")
+    
+    // Method implementation:
+    instructions( 
+        // Filter 1
+        fieldAccess(
+            definingClass = "this",
+            type = "Ljava/util/Map;"
+        ),
+
+        // Filter 2 
+        string("showBannerAds"),
+      
+        // Filter 3 
+        methodCall(
+            definingClass = "Ljava/lang/String;",
+            name = "equals",
+        ),
+
+        // Filter 4
+        // maxAfter = 0 means this must match immediately after the last filter.
+        opcode(Opcode.MOVE_RESULT, maxAfter = 0),
+
+        // Filter 5
+        literal(1337),
+        
+        // Filter 6 
+        opcode(Opcode.IF_EQ),
+    )
+    custom { method, classDef ->
+        classDef.type == "Lcom/some/app/ads/AdsLoader;"
     }
 }
 ```
 
-Using that fingerprint, this method can be matched uniquely from all other methods.
+  Notice the instruction filters do not declare every instruction in the target method,
+  and between each filter can exist 0 or more other instructions.  Instruction filters
+  must be declared in the same order as the instructions appear in the target method.
+
+  If the distance between each instruction declaration can be approximated,
+  then the `maxAfter` parameter can be used to restrict the instruction match to
+  a maximum distance from the last instruction.  A value of 0 for the first instruction filter
+  means the filter must be the first instruction of the target method.
+
+  If a single instruction varies slightly between different app targets but otherwise the fingerprint
+  is still the same, the `anyInstruction()` wrapper can be used to specify variations of the
+  same instruction.  Such as:
+  `anyInstruction(string("string in early app target"), string("updated string in latest app target"))`
+
+  If a method cannot be uniquely identified using the built in filters, but a fixed
+  pattern of opcodes can identify the method, then the opcode pattern can be
+  defined using the fingerprint `opcodes()` declaration.  Opcode patterns do not
+  allow variable spacing between each opcode, and all opcodes all must appear exactly as declared.
```

Comment:
Line breaks should be done at the default width IntelliJ has set up. Also ideally you break at sentence fragments and not arbitrarily.